src/server.c

```diff
@@ -201,6 +201,39 @@ mstime_t mstime(void) {
     return ustime()/1000;
 }
 
+/* Return the command time snapshot in milliseconds.
+ * The time the command started is the logical time it runs,
+ * and all the time readings during the execution time should
+ * reflect the same time.
+ * More details can be found in the comments below. */
+mstime_t commandTimeSnapshot(void) {
+    mstime_t now;
+
+    /* If we are in the context of a Lua script, we pretend that time is
+     * blocked to when the Lua script started. This way a key can expire
+     * only the first time it is accessed and not in the middle of the
+     * script execution, making propagation to slaves / AOF consistent.
+     * See issue #1525 on Github for more information. */
+    if (server.script_caller) {
+        now = scriptTimeSnapshot();
+    }
+    /* If we are in the middle of a command execution, we still want to use
+     * a reference time that does not change: in that case we just use the
+     * cached time, that we update before each call in the call() function.
+     * This way we avoid that commands such as RPOPLPUSH or similar, that
+     * may re-open the same key multiple times, can invalidate an already
+     * open object in a next call, if the next call will see the key expired,
+     * while the first did not. */
+    else if (server.fixed_time_expire > 0) {
+        now = server.mstime;
+    }
+    /* For the other cases, we want to use the most fresh time we have. */
+    else {
+        now = mstime();
+    }
```

Comment:
thinking about this more, i guess the only place where this `mstime()` call will kick in if for a module calling a RM_Call from a timer event or alike.
in which case i suppose using the cached time would be ok too.
@MeirShpilraien @guybe7 @soloestoy can you think of anything else?

in which case we can maybe completely drop the fixed_time_expire variable, and change this function to simply be:
```c
    if (server.script_caller) {
        return scriptTimeSnapshot();
    } else {
        return server.mstime;
    }
```

if we don't wanna do that, then i think we should at least rename the `fixed_time_expire` since it's no longer about expiry.

Reply:
if we think it's no longer about expiry, then we have much more cases need changes, such as the `time` command.

Reply:
i was referring to `fixed_time_expire` (which is no longer about expiry and should be renamed).
and yes, this PR aims to freeze time for everything, not just expiry.
and you're right that we missed the TIME command, but i'm not sure there are really may others that needs a change. what else?

maybe for modules.
i'm not certain about RM_Milliseconds (there's RM_CachedMicroseconds for that).
maybe RM_GetExpire and alike should probably get changed (@enjoy-binbin did you intentionally only scan commands and skip module.c?)

Reply:
i did see RM_Milliseconds and RM_CachedMicroseconds, and then i opened a question in https://github.com/redis/redis/pull/11016#pullrequestreview-1095644480, something about create a new RM_CachedMilliseconds. 

but i guess i missed RM_GetExpire and alike (I'm not sensitive enough to modules... i did recall i check these, but somehow i skipped it, `RM_GetExpire`, `RM_SetExpire` and `moduleBlockClient` should get changed)

how should we do about the TIME command? the microseconds part...

```
addReplyBulkLongLong(c,tv.tv_sec);
addReplyBulkLongLong(c,tv.tv_usec);
addReplyBulkLongLong(c,server.mstime/1000);
addReplyBulkLongLong(c, server.ustime - (server.ustime/1000000*1000000));

return:
1) "1663947631"
2) "489483"
3) "1663947631"
4) "489483"

addReplyBulkLongLong(c,tv.tv_sec);
addReplyBulkLongLong(c,tv.tv_usec);
addReplyBulkLongLong(c,commandTimeSnapshot()/1000);
addReplyBulkLongLong(c, commandTimeSnapshot() * 1000 - (commandTimeSnapshot() * 1000 / 1000000 * 1000000));

return: 
1) "1663947785"
2) "619490"
3) "1663947785"
4) "619000"
```

Reply:
what about
```
addReplyBulkLongLong(c,server.unixtime);
addReplyBulkLongLong(c,server.ustime-server.unixtime*1000000);
```

Reply:
generally speaking, when we want to sample time we should always use a time snapshot (so yes, let's get rid of `fixed_time_expire`)
we need to make sure that there are few calls to `server.mstime()` or alike, only in places where we actually need a fresh copy of the time (i think the rule of thumb is to use fresh-time when we're dealing with timeouts: for example the active-expire loop early-exit or when blocking a client)